### PR TITLE
Clusters Got Broken Up Unexpectedly

### PR DESCRIPTION
**Pull request by td928**
two reviewer preferred one required

address #352 which some clusters pretty clearly have spatial overlaps 

## sql/_project_record_ids.sql
there are only two main 

### ST_Contains(a.geom, b.intersect_geom)
this additional condition ensures that when the intersection geometry is entirely within the cluster geometries themselves they will still be grouped up to be in clusters. See below for the comparisons between the clusters produced from before and after the condition is added. 
[clusters_comparision.csv](https://github.com/NYCPlanning/db-knownprojects/files/10025822/clusters_comparision.csv)

### cluster id
The cluster id is added in two places to ensure that first the `ST_Intersections` would only be performed between projects that were first already identified by `dbscan`. This is implemented after some behaviors from the query were observed that when it is run on a single cluster it would behave correctly but when more records (multiple clusters) were brought in to process at once then. My intuition is that when `ST_Intersectiosn` were between project outside their previously identified the cluster by `dbscan` it created extra geometries that causes to `ST_Overlap` and `ST_Contains` to be confused and breaking up some clusters. Again, spot checking the clusters comparisons before and after should be illuminating in that which of the clusters should exist but didn't in the previous implementation.

### CODE REVIEW

**Review by AmandaDoyle — COMMENTED** [review]
I approve on the condition Max also approves it as well.  I confirmed that `ST_Overlaps` drops polygons that fall completely within another geometry and adding the `ST_Contains` clause, and am comfortable with the `WHERE (ST_Overlaps(a.geom, b.intersect_geom) or ST_Contains(a.geom, b.intersect_geom))` clause.  I'd like a second set of eyes on the `	AND a.id = b.id` clause to ensure that by only spatially joining records with the same id from the ST_ClusterDBSCAN does not exclude records that we want to include.  Based on my spot checks I do not see an issue; however, I'd like someone else to spot check and when we produce a full output for Housing team this is something we'll want to double check.

File: sql/_project_record_ids.sql [hunk 1]
```diff
@@ -34,19 +34,22 @@ project_record_join AS (
...
 	AND b.no_geom = 0
+	AND a.id = b.id
+	GROUP BY a.id
 )
 SELECT
 	array_agg(a.record_id) as project_record_ids
-INTO _project_record_ids
 FROM project_record_join a, all_intersections b
-WHERE ST_OVERLAPS(a.geom, b.intersect_geom)
+WHERE (ST_Overlaps(a.geom, b.intersect_geom) or ST_Contains(a.geom, b.intersect_geom))
```
Comment (mbh329) [inline sql/_project_record_ids.sql:51]: this is nitpicky but the "or" should be uppercase to follow the convention in this SQL file

File: sql/_project_record_ids.sql [hunk 2]
```diff
@@ -34,19 +34,22 @@ project_record_join AS (
...
 	WHERE a.record_id < b.record_id
 	AND a.records_in_project > 1
 	AND b.records_in_project > 1
 	AND a.no_geom = 0
 	AND b.no_geom = 0
+	AND a.id = b.id
+	GROUP BY a.id
 )
 SELECT
 	array_agg(a.record_id) as project_record_ids
```
Comment (mbh329) [inline sql/_project_record_ids.sql:49]: Also nitpicky but the "as" should be uppercase, I know we don't have a proper formatting tool for postgres...something we should think about

**Comment by mbh329** [comment]
The output looks good to me - I am a bit confused about what kind of spot checks I should be doing though

**Comment by mbh329** [comment]
@td928 the build is failing on my local, I think it might have to do with the edc_id in the edc_dcp_input file, I think you need to change line 86 in the combine.sql to b.edc_id::numeric. There were changes to dev that might need to be merged into this 352-ST_UNION-NULL

**Comment by td928** [comment]
> @td928 the build is failing on my local, I think it might have to do with the edc_id in the edc_dcp_input file, I think you need to change line 86 in the combine.sql to b.edc_id::numeric. There were changes to dev that might need to be merged into this 352-ST_UNION-NULL

great catch! Hopefully the merge from dev fix it. @mbh329

File: sql/_project_record_ids.sql [hunk 3]
```diff
@@ -27,26 +27,29 @@ dbscan AS (
...
-		ST_UNION(ST_INTERSECTION(a.geom, b.geom)) as intersect_geom
+	SELECT
+		ST_UNION(ST_INTERSECTION(a.geom, b.geom)) AS intersect_geom
 	FROM  project_record_join a, project_record_join b
 	WHERE a.record_id < b.record_id
 	AND a.records_in_project > 1
 	AND b.records_in_project > 1
 	AND a.no_geom = 0
 	AND b.no_geom = 0
+	AND a.id = b.id
```
Comment (mbh329) [inline sql/_project_record_ids.sql:45]: This line is breaking the build - why are there two project_record_join tables (a,b)? I think that is the issue but don't fully understand the logic behind having an `a` and `b` of the same table

**Comment by mbh329** [comment]
Something is going on with the `project_record_ids` table, can debug on Friday

File: sql/_project_record_ids.sql [hunk 4]
```diff
@@ -27,26 +27,30 @@ dbscan AS (
...
+	array_agg(a.record_id) AS project_record_ids
 FROM project_record_join a, all_intersections b
-WHERE ST_OVERLAPS(a.geom, b.intersect_geom)
+WHERE (ST_Overlaps(a.geom, b.intersect_geom) OR ST_Contains(a.geom, b.intersect_geom))
 AND a.id IS NOT NULL
-GROUP BY a.id;
\ No newline at end of file
+AND a.id = b.id
+GROUP BY a.id
+;
```
Comment (mbh329) [inline sql/_project_record_ids.sql:56]: To stop this table from printing to the console, this statement needs to be surrounded by parenthesis

**Comment by mbh329** [comment]
The build was successful but the `projecT_record_ids` on my local build don't always match the `new_cluster_ids` in the cluster_comparison table - how concerned should we be about that?

**Review by mbh329 — APPROVED** [review]
Great - thanks for going over this, it was super helpful to have a version to compare once you fixed the action

File: sql/_project_record_ids.sql [hunk 5]
```diff
@@ -27,26 +27,31 @@ dbscan AS (
...
+	AND a.id = b.id
+	GROUP BY a.id
 )
 SELECT
-	array_agg(a.record_id) as project_record_ids
+	array_agg(a.record_id) AS project_record_ids
 INTO _project_record_ids
 FROM project_record_join a, all_intersections b
-WHERE ST_OVERLAPS(a.geom, b.intersect_geom)
+WHERE (ST_Overlaps(a.geom, b.intersect_geom) OR ST_Contains(a.geom, b.intersect_geom))
```
Comment (mbh329) [inline sql/_project_record_ids.sql:53]: Still need to add the inverse logic to incorporate the edgecase we looked at

File: sql/_project_record_ids.sql [hunk 6]
```diff
@@ -27,26 +27,31 @@ dbscan AS (
...
+	AND a.id = b.id
+	GROUP BY a.id
 )
 SELECT
-	array_agg(a.record_id) as project_record_ids
+	array_agg(a.record_id) AS project_record_ids
 INTO _project_record_ids
 FROM project_record_join a, all_intersections b
-WHERE ST_OVERLAPS(a.geom, b.intersect_geom)
+WHERE (ST_Overlaps(a.geom, b.intersect_geom) OR ST_Contains(a.geom, b.intersect_geom) OR ST_Within(a.geom, b.intersect_geom))
```
Comment (mbh329) [inline sql/_project_record_ids.sql:53]: To properly identify different agency projects (polygon) as clusters, we utilize three postgis functions to cover different type of edge cases:
1. `St.Overlaps` - "Two geometries overlap if they have the same dimension, each has at least one point not shared by the other (or equivalently neither covers the other), and the intersection of their interiors has the same dimension" For our use case, `st.overlaps` is meant to include polygons in which their permiter crosses over into another polygon
st_overlaps03

2. `St_Contains `- "Returns TRUE if geometry B is completely inside geometry A. A contains B if and only if no points of B lie in the exterior of A, and at least one point of the interior of B lies in the interior of A. A subtlety of the definition is that a geometry does not contain things in its boundary. Thus polygons and lines do not contain lines and points lying in their boundary." For KPDB, we want to capture agency projects (polygons) that boundary is fully "contained" within another projects as a cluster. 
st_contains04
3. `ST_WIthin` - "Returns TRUE if geometry A is completely inside geometry B. For this function to make sense, the source geometries must both be of the same coordinate projection, having the same SRID. It is a given that if ST_Within(A,B) is true and ST_Within(B,A) is true, then the two geometries are considered spatially equal. ST_Within is the inverse of [ST_Contains](https://postgis.net/docs/ST_Contains.html). So, ST_Within(A,B) = ST_Contains(B,A)." For KPDB, we want to capture records that have the exact same geometry which is why we apply the inverse logic to st_contains(a.geom, b.intersect_geom) which would be st_contains(b.intersect_geom, a.geom)

Comment (mbh329) [reply sql/_project_record_ids.sql:53]: I think flushing out the comment for how we apply the spatial logic will be enough to approve this PR

**Comment by mbh329** [comment]
The output looks good! I think this is ready to be merged

**Pull request merged**